circle/src/util.rs

```diff
@@ -0,0 +1,177 @@
+    let mut cur = point.real();
+    for _ in 0..(log_n - 1) {
+        for i in 0..basis.len() {
+            basis.push(basis[i] * cur);
+        }
+        cur = F::two() * cur.square() - F::one();
+    }
+
+    // Bit reverse, and compute the second half of the array,
+    // which is just each element of the first half times y
+    reverse_slice_index_bits(&mut basis);
+    for i in 0..basis.len() {
+        basis.push(basis[i] * point.imag());
+    }
+
+    // Negate each element each time the binary representation of its index has a pair of adjacent ones,
+    // or equivalently, if the number of adjacent ones is odd.
+    // This comes from a peculiarity in how we compute the CFFT:
+    // The butterfly zips the first half of the domain with the second half reversed, because that maps each point
+    // to its involution. After each layer, the second half is still in reverse order, so we should use the twiddles
+    // in reverse order as well, but we ignore that and use the same twiddles for both halves.
+    // Using t(g^(N-k)) instead of t(g^k) just adds a negative sign. It turns out the number of negations is the number
+    // of adjacent ones in the index.
+    for (i, val) in basis.iter_mut().enumerate() {
+        let num_adjacent_ones = (i & (i >> 1)).count_ones();
+        if num_adjacent_ones % 2 == 1 {
+            *val = -*val;
+        }
+    }
+
+    basis
+}
+
+#[cfg(test)]
+pub(crate) fn eval_circle_polys<F: ComplexExtendable>(
+    coeffs: &RowMajorMatrix<F>,
+    point: Complex<F>,
+) -> Vec<F> {
+    use itertools::izip;
+    use p3_matrix::Matrix;
+
+    let log_n = log2_strict_usize(coeffs.height());
+    let mut accs = vec![F::zero(); coeffs.width()];
+    for (row, basis) in coeffs.rows().zip(circle_basis(point, log_n)) {
+        for (acc, coeff) in izip!(&mut accs, row) {
+            *acc += *coeff * basis;
+        }
+    }
+    accs
+}
+
+/// Circle STARKs, Section 3, Lemma 1: (page 4 of the first revision PDF)
+/// (x, y) = ((1-t^2)/(1+t^2), 2t/(1+t^2))
+/// t = ±i goes to (-1, 0)
+pub(crate) fn univariate_to_point<F: Field>(t: F) -> Complex<F> {
+    let t2 = t.square();
+    if let Some(inv_denom) = (F::one() + t2).try_inverse() {
+        Complex::new((F::one() - t2) * inv_denom, (F::two() * t) * inv_denom)
+    } else {
+        Complex::new_real(F::neg_one())
+    }
+}
+
+/// Circle STARKs, Section 3, Lemma 1: (page 4 of the first revision PDF)
+/// t = y / (x + 1)
+/// If F has i, this should return that instead, but we don't have access..
+pub(crate) fn point_to_univariate<F: Field>(p: Complex<F>) -> Option<F> {
+    p.imag().try_div(p.real() + F::one())
+}
+
+/// Formula for the group operation in univariate coordinates
+/// Circle STARKs, Section 3.1, Remark 4: (page 5 of the first revision PDF)
+/// same as above, this *could* handle point at infinity if we had Field::try_sqrt
+#[allow(unused)]
+pub(crate) fn rotate_univariate<F: Field, EF: ExtensionField<F>>(t1: EF, t2: F) -> Option<EF> {
```

Comment:
Do we really need the univariate representation of the circle group law?

Reply:
No, it's unused, I left it as it may be useful in the future